geopandas/io/parquet.py

```diff
@@ -0,0 +1,285 @@
+        ``os.PathLike``.
+
+        By file-like object, we refer to objects with a ``read()`` method,
+        such as a file handler (e.g. via builtin ``open`` function)
+        or ``StringIO``.
+    columns : list-like of strings, default=None
+        If not None, only these columns will be read from the file.  If
+        the primary geometry column is not included, the first secondary
+        geometry read from the file will be set as the geometry column
+        of the returned GeoDataFrame.  If no geometry columns are present,
+        a ``ValueError`` will be raised.
+    **kwargs
+        Any additional kwargs are passed to the engine.
+
+    Returns
+    -------
+    GeoDataFrame
+    """
+
+    import_optional_dependency(
+        "pyarrow", extra="pyarrow is required for parquet support."
+    )
+    from pyarrow import parquet
+
+    path, _, _, should_close = get_filepath_or_buffer(path)
+
+    kwargs["use_pandas_metadata"] = True
+    table = parquet.read_table(path, columns=columns, **kwargs)
+
+    df = table.to_pandas()
+
+    if should_close:
+        try:
+            path.close()
+        except:  # noqa: flake8
+            pass
+
+    metadata = None
+
+    try:
+        metadata = table.schema.metadata
+        if metadata is not None:
+            metadata = _decode_metadata(metadata.get(b"geo", b""))
+
+    except (TypeError, json.decoder.JSONDecodeError):
+        raise ValueError("Missing or malformed geo metadata in parquet file")
+
+    if metadata is None:
+        raise ValueError("Parquet file is missing required metadata 'geo' entry")
+
+    # Validate that required keys are present
+    required_keys = ("primary_column", "columns")
+    for key in required_keys:
+        if key not in metadata:
+            raise ValueError(
+                f"""'geo' metadata in parquet file is missing required key:
+                '{key}'"""
+            )
+
+    # Find all geometry columns that were read from the file.  May
+    # be a subset if 'columns' parameter is used.
+    geometry_columns = df.columns.intersection(metadata["columns"])
+
+    if not len(geometry_columns):
+        raise ValueError(
+            """No geometry columns are included in the columns read from
+            the parquet file.  To read this file without geometry columns,
+            use pandas.read_parquet() instead."""
+        )
+
+    column_metadata = metadata["columns"]
+
+    # Validate that geometry columns have required metadata and values
+    required_col_keys = ("crs", "encoding")
+    for col in geometry_columns:
+        for key in required_col_keys:
+            if key not in column_metadata[col]:
+                raise ValueError(
+                    f"""'geo' metadata in parquet file is missing required key
+                    {key} for column '{col}'"""
+                )
+
+        if column_metadata[col]["encoding"] != "WKB":
+            raise ValueError("Only WKB geometry encoding is supported")
```

Comment:
Maybe also move eg the validation of the metadata in a helper function? (if that makes sense to you, just a suggestion!)